File: Ruby/Civitas/pkg/Civitas-0.0.1/lib/vista_textual.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

# encoding: utf-8
require_relative 'operaciones_juego'
require_relative 'operacion_inmobiliaria'
require_relative 'gestiones_inmobiliarias'
require_relative 'jugador'
require_relative 'diario'
require_relative 'civitas_juego'
require_relative 'respuestas'
require_relative 'titulo_propiedad'
require 'io/console'

module Civitas
  class Vista_textual
    def mostrar_estado(estado)
      puts estado
    end
    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end

    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-#{l}"
        index += 1
      }
      opcion = lee_entero(lista.length,
                          "\n"+tab+" Elige una opcion: ",
                          tab+" Valor erroneo")
      return opcion
    end
    
    def set_civitas_juego(civitas)
         @juegoModel=civitas
         actualizar_vista
    end

    def actualizar_vista
      puts "#{@juegoModel.info_jugador_texto}
            \nCasilla -> #{@juegoModel.get_casilla_actual.to_s}"
    end
    
    def comprar
      respuestas=[Respuestas::SI, Respuestas::NO]
      opcion=menu(" Desea comprar la casilla actual?",["SI", "NO"])
      
      return(respuestas[opcion])
    end
    
    def salir_carcel
      opcion=menu("\nElige la forma para intentar salir de la carcel. \n", SalidasCarcel.lista_salidascarcel)
      salir=SalidasCarcel.lista_salidascarcel[opcion]
      return salir
    end

    def gestionar
      gestiones=[GestionesInmobiliarias::VENDER, GestionesInmobiliarias::HIPOTECAR, GestionesInmobiliarias::CANCELAR_HIPOTECA,
                 GestionesInmobiliarias::CONSTRUIR_CASA, GestionesInmobiliarias::CONSTRUIR_HOTEL, GestionesInmobiliarias::TERMINAR]
               
      opcion=menu("\nSeleccione gestion", gestiones)
      @iGestion=gestiones[opcion]
      
      calles=Array.new
      
      for i in (0..@juegoModel.get_jugador_actual.get_propiedades.length-1)
        calles<<@juegoModel.get_jugador_actual.get_propiedades.at(i).get_nombre
      end
      
      if(@iGestion!=Civitas::GestionesInmobiliarias::TERMINAR)
        @iPropiedad=menu("\nSobre que casilla desea realizar la gestion?", calles)
      end
    end

    def get_gestion
      gest=@iGestion
      
      return gest
    end

    def get_propiedad
      prop=@iPropiedad
      
      return prop
    end

    def mostrar_siguiente_operacion(operacion)
      case operacion
      when Civitas::Operaciones_juego::AVANZAR
        puts "Proxima operacion: Avanzar\n"
      when Civitas::Operaciones_juego::COMPRAR 
        puts "Proxima operacion: Comprar\n"
      when Civitas::Operaciones_juego::GESTIONAR
        puts "Proxima operacion: Gestionar\n"
      when Civitas::Operaciones_juego::PASAR_TURNO
        puts "Proxima operacion: Pasar turno\n"
      when Civitas::Operaciones_juego::SALIR_CARCEL
        puts "Proxima operacion: Salir de la carcel\n"
      end
    end

    def mostrar_eventos
      puts "\nDIARIO:"
      
      while(Diario.instance.eventos_pendientes)
        puts "\n-#{Diario.instance.leer_evento}"
      end
    end
  end
end

File: Java/P4/Civitas/src/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.civitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.Jugador;
import civitas.tituloPropiedad;
import civitas.Respuestas;

class VistaTextual {
  
  civitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print (" Pulsa una tecla ");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la cárcel",
      new ArrayList<> (Arrays.asList("Pagando", "Tirando")));
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar(){
    Respuestas compra;
    int op=menu ("¿Desea comprar?",
    new ArrayList<> (Arrays.asList("NO","SI")));
    compra=Respuestas.values()[op];
    return compra;
  }

  void gestionar(){
      int opcion=menu("¿Qué número de gestión inmobiliaria desea realizar?", new ArrayList<>(Arrays.asList("Vender", "Hipotecar", "Cancelatr Hipoteca", "Construir casa", "Construir hotel", "Terminar")));
      iGestion=opcion;
      
      ArrayList<String> calles=new ArrayList<String>();
      
      for(int i=0; i<juegoModel.getJugadorActual().getPropiedades().size(); i++){
          calles.add(juegoModel.getJugadorActual().getPropiedades().get(i).getNombre());
      }
      
      iPropiedad=menu("¿Sobre qué casilla desea realiza la operación?", calles);   
  }
  
  public int getGestion(){
      return iGestion;
  }
  
  public int getPropiedad(){
      return iPropiedad;
  }
    
  void mostrarSiguienteOperacion(OperacionesJuego operacion) {
  System.out.print("Operación siguiente: "+operacion);}


  void mostrarEventos() {
      if(Diario.getInstance().eventosPendientes()){
          System.out.print(Diario.getInstance().leerEvento());
      }
  }
  
  public void setCivitasJuego(civitasJuego civitas){ 
        juegoModel=civitas;
        this.actualizarVista();
  }
  
  void actualizarVista(){
      System.out.print(juegoModel.infoJugadorTexto());
  } 
}